neps/nep-0591.md

```diff
@@ -0,0 +1,207 @@
+Status: Draft
+DiscussionsTo: https://github.com/nearprotocol/neps/pull/591
+Type: Protocol
+Replaces: 491
+Version: 1.0.0
+Created: 2025-02-11
+LastUpdated: 2025-02-11
+---
+
+## Summary
+
+This proposal introduces global contracts, a new mechanism that allows smart contracts to be deployed once and reused by any account without incurring high storage costs.
+
+Currently, deploying the same contract multiple times on different accounts leads to significant storage fees.
+Global contracts solve this by making contract code available globally, allowing multiple accounts to reference it instead of storing their own copies.
+
+Rather than requiring full storage costs for each deployment, accounts can simply link to an existing global contract, reducing redundancy and improving scalability. This approach optimizes storage, lowers costs, and ensures efficient contract distribution across the network.
+
+## Motivation
+
+A common use case on NEAR is to deploy the same smart contract many times on many different accounts. For example, a multisig contract is a frequently deployed contract.
+However, today each time such a contract is deployed, a user has to pay for its storage and the cost is quite high. For a 300kb contract the cost is 3N.
+
+With the advent of chain signatures, the smart contract wallet use case will become more ubiquitous.
+As a result, it is very desirable to be able to reuse already deployed contract without having to pay for the storage cost again.
+
+Additionally global contracts cover the underlying use case for [NEP-491](https://github.com/near/NEPs/pull/491): https://github.com/near/nearcore/pull/12818.
+
+## Specification
+
+Global contract can be deployed in 2 ways: either by its hash or by owner account id.
+Contracts deployed by hash are effectively immutable and cannot be updated.
+When deployed by account id the owner can redeploy the contract updating it for all its users.
+Users can use contracts deployed by hash if they prefer having control over contract updates. In order to update the contract user would have to explicitly switch to a different version of the contract deployed under a different hash.
+Contracts deployed by account id should be used when user trusts contract developers to update the contract for them. For example if user accounts are created specifically for some application to be onchain.
+
+We introduce new receipt action for deploying global contracts:
+
+```rust
+struct DeployGlobalContractAction {
+    code: Vec<u8>,
+    deploy_mode: GlobalContractDeployMode,
+}
+
+enum GlobalContractDeployMode {
+    /// Contract is deployed under its code hash.
+    /// Users will be able reference it by that hash.
+    /// This effectively makes the contract immutable.
+    CodeHash,
+    /// Contract is deployed under the owner account id.
+    /// Users will be able reference it by that account id.
+    /// This allows the owner to update the contract for all its users.
+    AccountId,
```

Comment:
Does deploying a global contract by account id deploy said contract to the owner account as well?

Reply:
nope, it does not do that